### PR TITLE
Beta

### PR DESCRIPTION
**Pull request by markhollingworth-worthit**
Details

- Rename /restartbots to /resumebots
- Fixed error when {market}.json file might not be complete (bot stopped).
- Add /cleandata to remove any {market}.json file which are not complete (bot stopped)
- Add not to create telegram_data folder when running sims
- Add /startscanner to TG Bot to scan marks and start bots (hourly schedule set in config)
- Add /stopscanner to TG Bot to stop scanning markets
- Add /addexception to TG Bot to add pair to exceptions list so wont be started by scanner process
- Add /removeexception to TG Bot to remove a pair from exceptions list
- Add multiple market scanning to scanner.py
- Add exception list so that scanner.py doesnt pick a pair your dont want traded

## Type of change

Please make your selection.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change requires a new dependency
- [ ] Code Maintainability / comments
- [x] Code Refactoring / future-proofing

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added pytest unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: models/AppState.py [hunk 1]
```diff
@@ -207,8 +207,17 @@ def getLastOrder(self):
...
 
-        base = float(self.account.getBalance(self.app.getBaseCurrency()))
-        quote = float(self.account.getBalance(self.app.getQuoteCurrency()))
+        ac = self.account.getBalance()
+
+        df_base = ac[ac["currency"] == self.app.getBaseCurrency()]["available"]
+        base = 0.0 if len(df_base) == 0 else float(ac[ac["currency"] == self.app.getBaseCurrency()]["available"].values[0])
+
+        df_quote = ac[ac["currency"] == self.app.getQuoteCurrency()]["available"]
+        quote = 0.0 if len(df_quote) == 0 else float(ac[ac["currency"] == self.app.getQuoteCurrency()]["available"].values[0])
```
Comment (edrickrenan) [inline models/AppState.py:216]: `quote = 0.0 if len(df_quote) == 0 else float(df_quote.values[0])`?

File: models/AppState.py [hunk 2]
```diff
@@ -207,8 +207,17 @@ def getLastOrder(self):
             self.last_action = "SELL"
             return
 
-        base = float(self.account.getBalance(self.app.getBaseCurrency()))
-        quote = float(self.account.getBalance(self.app.getQuoteCurrency()))
+        ac = self.account.getBalance()
+
+        df_base = ac[ac["currency"] == self.app.getBaseCurrency()]["available"]
+        base = 0.0 if len(df_base) == 0 else float(ac[ac["currency"] == self.app.getBaseCurrency()]["available"].values[0])
```
Comment (edrickrenan) [inline models/AppState.py:213]: `base = 0.0 if len(df_base) == 0 else float(df_base.values[0])`?

File: models/exchange/coinbase_pro/api.py [hunk 3]
```diff
@@ -35,6 +36,16 @@ def _isMarketValid(self, market: str) -> bool:
             return True
         return False
 
+    def to_coinbasepro_granularity(self, granularity) -> str:
```
Comment (edrickrenan) [inline models/exchange/coinbase_pro/api.py:39]: Unused. Can be deleted

File: models/exchange/coinbase_pro/api.py [hunk 4]
```diff
@@ -1153,18 +1167,19 @@ def on_message(self, msg):
...
-            elif self.granularity == 900:
-                df["candle"] = df["date"].dt.floor(freq="15T")
-            elif self.granularity == 3600:
-                df["candle"] = df["date"].dt.floor(freq="1H")
-            elif self.granularity == 21600:
-                df["candle"] = df["date"].dt.floor(freq="6H")
-            elif self.granularity == 86400:
-                df["candle"] = df["date"].dt.floor(freq="1D")
+            df["candle"] = df["date"].dt.floor(freq=self.granularity.frequency)
+            # if self.granularity.frequency == 60:
```
Comment (edrickrenan) [inline models/exchange/coinbase_pro/api.py:1171]: The commented lines can be removed.

File: models/exchange/kucoin/api.py [hunk 5]
```diff
@@ -41,6 +41,18 @@ def _isMarketValid(self, market: str) -> bool:
             return True
         return False
 
+    def to_kucoin_granularity(self, granularity) -> str:
```
Comment (edrickrenan) [inline models/exchange/kucoin/api.py:44]: can be removed

File: scanner.json [hunk 6]
```diff
@@ -1,4 +1,8 @@
 {
-    "exchange": "coinbasepro",
-    "quote_currency": "GBP"
+    "coinbasepro" : {
+        "quote_currency": ["GBP"]
+    },
+    "binance" : {
+        "quote_currency": ["USDT"]
```
Comment (edrickrenan) [inline scanner.json:6]: Maybe using `BUSD` by default would be better?

File: scanner.py [hunk 7]
```diff
@@ -1,90 +1,108 @@
...
-    api = KPublicAPI()
-else:
-    raise ValueError(f"Invalid exchange: {config['exchange']}")
+for exchange in config:
+    ex = Exchange(exchange)
+    app = PyCryptoBot(exchange=ex)
+    for quote in config[ex.value]["quote_currency"]:
+        if ex == Exchange.BINANCE:
+            api = BPublicAPI()
+            # GRANULARITY = Granularity.convert_to_enum(3600).to_short
```
Comment (edrickrenan) [inline scanner.py:27]: can be removed

**Comment by whittlem** [comment]
Thanks @markhollingworth-worthit, there are some conflicts. Can you please resolve?

**Comment by Adawntoremember** [comment]
9a57e0b2320b804baccc4d5faeb99c9fa5bc7c64

**Pull request merged**